### PR TITLE
[CI,WIP, enhancement] add pytorch+gpu testing ci

### PR DESCRIPTION
**Pull request by icfaust**
## Description

_Add a comprehensive description of proposed changes_

_List associated issue number(s) if exist(s): #6 (for example)_

_Documentation PR (if needed): #1340 (for example)_

_Benchmarks PR (if needed): https://github.com/IntelPython/scikit-learn_bench/pull/155 (for example)_

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: deselected_tests.yaml [hunk 1]
```diff
@@ -448,3 +469,6 @@ gpu:
 
   # Introduced with RNG forest updates in oneDAL
   - ensemble/tests/test_voting.py::test_set_estimator_drop
+
+  # Deselection for public GPU runner
+  - model_selection/tests/test_validation.py::test_learning_curve_some_failing_fits_warning[42]
```
Comment (david-cortes-intel) [inline deselected_tests.yaml]: Does this deselect the test for one specific random seed? Does it pass for other seeds?

File: daal4py/sklearn/utils/validation.py [hunk 2]
```diff
@@ -72,21 +72,20 @@ def _assert_all_finite(
...
-                    msg_dtype=msg_dtype,
-                    estimator_name=estimator_name,
-                    input_name=input_name,
-                )
-            else:
-                _sklearn_assert_all_finite(X, allow_nan=allow_nan, msg_dtype=msg_dtype)
-            return
-
     is_df = is_DataFrame(X)
+    if not (is_df or isinstance(X, np.ndarray)) or X.size < 32768:
```
Comment (david-cortes-intel) [inline daal4py/sklearn/utils/validation.py:76]: Wouldn't this need the same pytorch workaround used in the tests for the size attribute?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -313,3 +317,175 @@ jobs:
...
+          echo "NUMPY_BUILD=$(python -m pip freeze | grep numpy)" >> "$GITHUB_ENV"
+          bash .ci/scripts/setup_sklearn.sh ${{ env.UXL_SKLEARNVERSION }}
+          pip install --upgrade -r requirements-test.txt
+          pip install $(python .ci/scripts/get_compatible_scipy_version.py ${{ env.UXL_SKLEARVERSION }}) pyyaml
+          pip list
+      - name: Download sklearnex wheel
+        uses: actions/download-artifact@v4
+        with:
+          name: sklearnex_build_${{ env.UXL_PYTHONVERSION }}
+      - name: Install PyTorch
```
Comment (david-cortes-intel) [inline .github/workflows/ci.yml:466]: Do you think this could make a check to verify that the install recognizes the GPU? Like in here:
https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html#check-availability-for-intel-gpu

**Comment by david-cortes-intel** [comment]
@icfaust Could the jobs be renamed to have 'torch' in the names instead of generic 'gpu'?

File: daal4py/sklearn/utils/validation.py (hunk 2 above)
Comment (icfaust) [reply daal4py/sklearn/utils/validation.py:76]: first part of the conditional should block that

Comment (david-cortes-intel) [reply daal4py/sklearn/utils/validation.py:76]: But the condition is 'if not', so a torch array would meet the first part (not df, not numpy) and the second one would then be evaluated on it.

Comment (icfaust) [reply daal4py/sklearn/utils/validation.py:76]: you are correct that the first condition will evaluate true and the following or command would not be evaluated. it would already know that sklearn must be used.

example
```python
if not (False or False) or 0/0:
    print("foobar")
```
should print 'foobar'

File: deselected_tests.yaml (hunk 1 above)
Comment (icfaust) [reply deselected_tests.yaml]: Good question, it looks like sklearn has a specific test fixutre for that value as its the ```global_random_seed``` (https://github.com/scikit-learn/scikit-learn/blob/1.4.X/sklearn/model_selection/tests/test_validation.py#2420) just as a remark, this is also being deselected in #2465 and some information on it is available there for from @ethanglaser

**Comment by icfaust** [comment]
> @icfaust Could the jobs be renamed to have 'torch' in the names instead of generic 'gpu'?

I'll make sure to add the frameworks to the title. I'd like to also still keep the `cpu` vs `gpu` because of how it impacts sklearn conformance testing.  Its the only piece currently running on GPU in this PR. Let me know what you think.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (icfaust) [reply .github/workflows/ci.yml:466]: Will do! Seems like a good thing to add

**Comment by david-cortes-intel** [comment]
> > @icfaust Could the jobs be renamed to have 'torch' in the names instead of generic 'gpu'?
> 
> I'll make sure to add the frameworks to the title. I'd like to also still keep the `cpu` vs `gpu` because of how it impacts sklearn conformance testing. Its the only piece currently running on GPU in this PR. Let me know what you think.

torch x [cpu, gpu] sounds good.

File: deselected_tests.yaml (hunk 1 above)
Comment (david-cortes-intel) [reply deselected_tests.yaml]: @icfaust But that other PR is not putting a specific seed to exclude in this test.

**Comment by david-cortes-intel** [comment]
@icfaust Is https://github.com/uxlfoundation/scikit-learn-intelex/pull/2465 meant to be merged before this?

File: deselected_tests.yaml (hunk 1 above)
Comment (icfaust) [reply deselected_tests.yaml]: Good question, both will deselect this test, as it is a single parameter single value test as they want it deterministic (specific details can be found here) https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/conftest.py#L318 where the global_random_seed default value is where it is defined.

**Comment by icfaust** [comment]
> @icfaust Is #2465 meant to be merged before this?

Another good question. It isn't a requirement and are independent of one another. They are related as they are both testing sklearnex on GPU vs sklearn 1.4 in CI for the first time.

File: deselected_tests.yaml (hunk 1 above)
Comment (david-cortes-intel) [reply deselected_tests.yaml]: Then how about leaving it deselected as a whole instead of parameterizing the deselection? Leaving it like this gives the impression that there's some specific kind of input that is problematic for the test, whereas it would still fail with any other input.

File: .ci/scripts/run_sklearn_tests.sh [hunk 4]
```diff
@@ -29,8 +29,11 @@ export DESELECT_FLAGS="--public ${DESELECT_FLAGS}"
 if [ -n "${SKLEARNEX_PREVIEW}" ]; then
     export DESELECT_FLAGS="--preview ${DESELECT_FLAGS}"
 fi
-export DESELECTED_TESTS=$(python ../.circleci/deselect_tests.py ../deselected_tests.yaml ${DESELECT_FLAGS})
```
Comment (ahuber21) [inline .ci/scripts/run_sklearn_tests.sh:32]: I know we've `export`ed before, but is this even necessary? Looks like we're just providing the list as command line argument. Seems somehow inconsistent to me.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -313,3 +317,177 @@ jobs:
           if "${{ steps.set-env.outputs.DPCFLAG }}"=="" set CPU=cpu
           set SKLEARNEX_PREVIEW=YES
           bash .ci/scripts/run_sklearn_tests.sh %CPU%
+
+  build_uxl:
+    if: github.repository == 'uxlfoundation/scikit-learn-intelex'
+    needs: onedal_nightly
+    name: LinuxNightly build Python${{ needs.onedal_nightly.outputs.uxl-python }}
+    runs-on: uxl-xlarge
+    timeout-minutes: 30
```
Comment (ahuber21) [inline .github/workflows/ci.yml:326]: How much headroom is 30 mins? I'd say we should give it at least 4x the expected runtime, and depending on how many cores we're using it could run some 10 or so minute?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -313,3 +317,177 @@ jobs:
...
+          path: .ci/env
+      - name: Set Environment Variables
+        id: set-env
+        run: |
+          # Disable SPMD testing
+          echo "NO_DIST=1" >> "$GITHUB_ENV"
+          # enable coverage report generation
+          echo "SKLEARNEX_GCOV=1" >> "$GITHUB_ENV"
+      - name: apt-get
+        run: sudo apt-get update && sudo apt-get install -y clang-format
```
Comment (ahuber21) [inline .github/workflows/ci.yml:363]: It is totally out of scope, but `clang-format` can be installed through pip. Maybe next time when you're playing around with the setup, experiment with adding `clang-format` to the dev requirements?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -313,3 +317,177 @@ jobs:
...
+        id: set-env
+        run: |
+          echo "NO_DIST=1" >> "$GITHUB_ENV"
+          # enable coverage report generation
+          echo "COVERAGE_RCFILE=$(readlink -f .coveragerc)" >> "$GITHUB_ENV"
+          echo "ONEDAL_PYTEST_FRAMEWORKS=${{ matrix.FRAMEWORKS }}" >> "$GITHUB_ENV"
+          # reduce GPU driver/runner related memory issues
+          echo "NEOReadDebugKeys=1" >> "$GITHUB_ENV"
+          echo "EnableRecoverablePageFaults=1" >> "$GITHUB_ENV"
+          echo "GpuFaultCheckThreshold=0" >> "$GITHUB_ENV"
```
Comment (ahuber21) [inline .github/workflows/ci.yml:443]: We could argue that those should be set in `run_sklearn_tests.sh`. Maybe that's neater?

File: onedal/_device_offload.py [hunk 8]
```diff
@@ -126,10 +127,29 @@ def _transfer_to_host(*data):
                 order=order,
             )
             has_usm_data = True
-        elif array_api and not _is_numpy_namespace(array_api):
-            # `copy`` param for the `asarray`` is not setted.
-            # The object is copied only if needed.
-            item = np.asarray(item)
+        elif not isinstance(item, np.ndarray) and (
```
Comment (ahuber21) [inline onedal/_device_offload.py:132]: Since you're adding so many new lines to an already long function with many branches, can I request that you refactor `_transfer_to_host()`. I'd suggest to put the logic for each case into a dedicated function and only set `has_usm_data` and `has_host_data` directly in the if statements.

File: onedal/datatypes/_data_conversion.py [hunk 9]
```diff
@@ -18,6 +18,9 @@
 
 from onedal import _default_backend as backend
 
+kDLCPU = backend.kDLCPU
```
Comment (ahuber21) [inline onedal/datatypes/_data_conversion.py]: Are those global symbols necessary?

File: sklearnex/utils/validation.py [hunk 10]
```diff
@@ -70,7 +71,14 @@ def _sklearnex_assert_all_finite(
     # size check is an initial match to daal4py for performance reasons, can be
     # optimized later
     xp, _ = get_namespace(X)
-    if X.size < 32768 or not _onedal_supported_format(X, xp):
+    # this try-catch is a PyTorch-specific fix, as Tensor.size is a function.
+    # The try-catch minimizes changes to most common code path (numpy arrays).
+    try:
+        too_small = X.size < 32768
+    except TypeError:
+        too_small = math.prod(X.shape) < 32768
```
Comment (ahuber21) [inline sklearnex/utils/validation.py]: Wouldn't this work?

```suggestion
    too_small = (math.prod(X.shape) if xp is torch else X.size) < 32768
```

**Review by ahuber21 — DISMISSED** [review]

File: deselected_tests.yaml (hunk 1 above)
Comment (icfaust) [reply deselected_tests.yaml]: Sounds good!

File: onedal/_device_offload.py (hunk 8 above)
Comment (icfaust) [reply onedal/_device_offload.py:132]: You and I are on the same page, done in the follow up to this PR: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2509 which will simplify the ```__sycl_usm_array_interface__``` branch, and then an yet unstarted PR will separate out the ```__dlpack_device__``` branch to a new file under ```onedal/datatypes/``` as that will coincide with new queue extraction added for ```from_data``` in ```onedal/utils/_sycl_queue_manager.py``` (as a follow-up to #2509).

File: sklearnex/utils/validation.py (hunk 10 above)
Comment (icfaust) [reply sklearnex/utils/validation.py]: Its definitely cleaner. The only problem is that we would have to have imported torch, which should be avoided unless absolutely necessary.  I could modify your suggestion to use array_api_compat's ```is_torch_array``` (https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py#L149). Let me know what you think.

File: onedal/datatypes/_data_conversion.py (hunk 9 above)
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py]: You highlight something that I did which wasn't so perfect.  The dlpack device data should be on CPU, the most sure way to do that is to check if its the kDLCPU enum.  This will only be used in _transfer_to_host, where the kDLOneAPI will be used for recognizing if a queue can be extracted.  I need to get it in ```onedal/_device_offload.py``` do you think I should import it from the backend there directly? I put it in ```_data_conversion``` since its a data conversion topic, but it required importing it through 2-3 spots to get it in _device_offload. Let me know what you think.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (icfaust) [reply .github/workflows/ci.yml:443]: The only worry is that it will only impact the driver settings for the sklearn conformance steps, will need to do it also for the sklearnex testing. We could move it to ```run_sklearn_tests.sh```, and in doing so add it to ```conda_recipe/run_test.sh``` and possibly ```conda_recipe/run_test.bat```. I am willing to do it, just let me know.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (icfaust) [reply .github/workflows/ci.yml:363]: Honestly that's a great idea, half the time when I am building a new environment I run head long into this issue at the daal4py build step.  Having a proper versioned clang-format from pip would be great to have in dev-dependencies to simplify my life, especially because we are already referencing the pip clang-format in the pre-commit step anyway.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (icfaust) [reply .github/workflows/ci.yml:326]: Never looked deeply into the value.  Last run was approx 10 min on the build step, so I guess its a 3x buffer, do you think I should bump to 40 min?

File: .ci/scripts/run_sklearn_tests.sh (hunk 4 above)
Comment (icfaust) [reply .ci/scripts/run_sklearn_tests.sh:32]: oh yeah, this stuff could be improved from the original design in #1199. Its re-acquired in the python script called in the shell script: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/.ci/scripts/run_sklearn_tests.py#L54 .  Truly should just be a command line argument into ```run_sklearn_tests.py```.  Would it be okay to assign to me as a follow up to re-jigger that aspect?

File: .github/workflows/ci.yml (hunk 5 above)
Comment (ahuber21) [reply .github/workflows/ci.yml:326]: I guess it's fine

File: onedal/_device_offload.py (hunk 8 above)
Comment (ahuber21) [reply onedal/_device_offload.py:132]: Cool.

File: onedal/datatypes/_data_conversion.py (hunk 9 above)
Comment (ahuber21) [reply onedal/datatypes/_data_conversion.py]: I would always import it straight from the original source. Let me know if that causes any issues.

File: sklearnex/utils/validation.py (hunk 10 above)
Comment (ahuber21) [reply sklearnex/utils/validation.py]: Yeah, so we do this?

```python
is_torch = is_torch_array(X)
too_small = (math.prod(X.shape) if is_torch else X.size) < 32768
```

File: .github/workflows/ci.yml (hunk 7 above)
Comment (ahuber21) [reply .github/workflows/ci.yml:443]: No strong opinion, it was just a thought. Let's leave it like this. Keeping multiple script files in sync isn't any better.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -313,3 +317,177 @@ jobs:
...
+        run: |
+          # This CI system yields oneAPI dependencies from the oneDAL repository
+          bash .ci/env/apt.sh dpcpp
+      - name: describe system
+        run: |
+          source /opt/intel/oneapi/setvars.sh
+          bash .ci/scripts/describe_system.sh
+      - name: Install test requirements
+        run: |
+          pip install -r dependencies-dev
```
Comment (ethanglaser) [inline .github/workflows/ci.yml]: I believe installation of dependencies-dev be skipped for test env

Comment (icfaust) [reply .github/workflows/ci.yml]: Oh interesting. You are absolutely right, if we have implemented requirements-test.txt properly, then dependencies-dev shouldn't be necessary.  I'll remove it and see what happens, if it works, now we would have a way of testing publicly that they work as intended separately (which we otherwise don't currently do).

File: .github/workflows/ci.yml (hunk 6 above)
Comment (icfaust) [reply .github/workflows/ci.yml:363]: Started in #2543

File: .github/workflows/ci.yml (hunk 11 above)
Comment (icfaust) [reply .github/workflows/ci.yml]: Done! looks like it works.

**Comment by david-cortes-intel** [comment]
> Something broke with the new coverage stats and need to investigate it.

@icfaust It's a general world-wide GCP outage.

File: sklearnex/utils/validation.py (hunk 10 above)
Comment (icfaust) [reply sklearnex/utils/validation.py]: Done

File: onedal/datatypes/_data_conversion.py (hunk 9 above)
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py]: Done

**Comment by icfaust** [comment]
Had to switch to just using ```math.prod(X.shape)``` just like sklearn due to private CI infrastructure issues.  This removes the array_api_compat dependency.
